File: src/DeviceRevert.cpp
// this file is covered by the  GNU LESSER GENERAL PUBLIC LICENSE Version 3 or later
// please see LICENSE.txt for more details and licensing issues
// copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

#include "DeviceRevert.h"
#include "OpticalDevice.h"

#include <cassert>

////////////////////////////////////////////////////////////////////////////////
//swap surfaces and change sign of curvature and ticks
bool DeviceRevert::revert(OpticalDevice* pDevice, int iFirstSurface, int iLastSurface)
{
	assert(pDevice != 0);
	OpticalDevice opdOld = *pDevice;

	for (int iSurf = iFirstSurface; iSurf <= iLastSurface; iSurf++)
	{
		Surface s2 = opdOld.surface(iSurf);
		
		//invert profile
		s2.set_radius_curvature( -s2.radius_curvature());
		s2.set_R4(-s2.R4());
		s2.set_R6(-s2.R6());
		s2.set_R8(-s2.R8());
		s2.set_R10(-s2.R10());

		pDevice->surface(iLastSurface - iSurf) = s2;
	}

	//permute all ticks but the last
	for (int iSurf = iFirstSurface; iSurf <= iLastSurface; iSurf++)
	{
		pDevice->set(iSurf, Z, opdOld.get(iLastSurface - iSurf - 1, Z));
	}
	//pDevice->set(iLastSurface, THICK, opdOld.get(iLastSurface, THICK));

    return true;
}
//////////////////////////////////////////////////////////////////////////////

File: python_astree_old/DeviceRevert.py
import math

def revert(pDevice, iFirstSurface, iLastSurface):
    assert pDevice is not None
    opdOld = pDevice.copy()

    for iSurf in range(iFirstSurface, iLastSurface+1):
        pDevice.set(iSurf, 'RADIUS_CURVATURE', -opdOld.get(iLastSurface - iSurf, 'RADIUS_CURVATURE'))
        pDevice.set(iSurf, 'R4', -opdOld.get(iLastSurface - iSurf, 'R4'))
        pDevice.set(iSurf, 'R6', -opdOld.get(iLastSurface - iSurf, 'R6'))
        pDevice.set(iSurf, 'R8', -opdOld.get(iLastSurface - iSurf, 'R8'))
        pDevice.set(iSurf, 'R10', -opdOld.get(iLastSurface - iSurf, 'R10'))
        pDevice.set(iSurf, 'CONIC', -opdOld.get(iLastSurface - iSurf, 'CONIC'))

        pDevice.set(iSurf, 'DIAMETER', opdOld.get(iLastSurface - iSurf, 'DIAMETER'))
        pDevice.set(iSurf, 'INNER_DIAMETER', opdOld.get(iLastSurface - iSurf, 'INNER_DIAMETER'))

    for iSurf in range(iFirstSurface+1, iLastSurface):
        pDevice.set(iSurf, 'Z', opdOld.get(iLastSurface, 'Z') - opdOld.get(iLastSurface - iSurf, 'Z') + opdOld.get(iFirstSurface, 'Z'))

    for iSurf in range(iFirstSurface, iLastSurface):
        pDevice.set_type(iSurf, opdOld.type(iLastSurface - iSurf - 1))

    return True